core/src/main/java/org/apache/brooklyn/core/catalog/internal/BundleUpgradeParser.java

```diff
@@ -0,0 +1,239 @@
+                throw new IllegalArgumentException("Must not be blank");
+            }
+            String[] parts = val.split(":");
+            if (parts.length > 2) {
+                throw new IllegalArgumentException("Identifier '"+val+"' has too many parts; max one ':' symbol");
+            }
+            if (parts.length == 1 || Strings.isBlank(parts[1])) {
+                throw new IllegalArgumentException("Identifier '"+val+"' must be of 'name:versionRange' syntax");
+            } else if (singleVersionIsOsgiRange || (parts[1].startsWith("(") || parts[1].startsWith("["))) {
+                return new VersionRangedName(parts[0], parts[1]);
+            } else {
+                return new VersionRangedName(parts[0], "["+parts[1]+","+parts[1]+"]");
+            }
+        }
+
+        public VersionRangedName(String name, VersionRange v) {
+            this.name = checkNotNull(name, "name").toString();
+            this.v = checkNotNull(v, "versionRange").toString();
+        }
+        
+        private VersionRangedName(String name, String v) {
+            this.name = checkNotNull(name, "name");
+            this.v = checkNotNull(v, "versionRange");
+        }
+        
+        @Override
+        public String toString() {
+            return name + ":" + v;
+        }
+        
+        public String toOsgiString() {
+            return name + ":" + getOsgiVersionRange();
+        }
+
+        public String getSymbolicName() {
+            return name;
+        }
+
+        public VersionRange getOsgiVersionRange() {
+            if (cachedOsgiVersionRange == null) {
+                cachedOsgiVersionRange = VersionRange.valueOf(BrooklynVersionSyntax.toValidOsgiVersionRange(v));
+            }
+            return cachedOsgiVersionRange;
+        }
+
+        public String getVersionString() {
+            return v;
+        }
+        
+        @Override
+        public int hashCode() {
+            return Objects.hashCode(name, v);
+        }
+
+        @Override
+        public boolean equals(Object other) {
+            if (!(other instanceof VersionRangedName)) {
+                return false;
+            }
+            VersionRangedName o = (VersionRangedName) other;
+            return Objects.equal(name, o.name) && Objects.equal(v, o.v);
+        }
+    }
+
+    public static CatalogUpgrades parseBundleManifestForCatalogUpgrades(Bundle bundle) {
+        Dictionary<String, String> headers = bundle.getHeaders();
+        String removedLegacyItemsHeader = headers.get(MANIFEST_HEADER_FORCE_REMOVE_LEGACY_ITEMS);
+        String removedBundlesHeader = headers.get(MANIFEST_HEADER_FORCE_REMOVE_BUNDLES);
+        List<VersionRangedName> removedLegacyItems = ImmutableList.of();
+        List<VersionRangedName> removedBundles = ImmutableList.of();
+        if (removedLegacyItemsHeader == null && removedBundlesHeader == null) {
+            return CatalogUpgrades.EMPTY;
+        }
+        if (removedLegacyItemsHeader != null) {
```

Comment:
if it were me i'd have `parseVersionRangedNameList(null)` return empty list and you can rewrite the entire method as

```
return CatalogUpgrades.builder()
                 .removedLegacyItems(parseVersionRangedNameList(headers.get(MANIFEST_HEADER_FORCE_REMOVE_LEGACY_ITEMS)))
                 .removedBundles(parseVersionRangedNameList(headers.get(MANIFEST_HEADER_FORCE_REMOVE_BUNDLES)))
                 .build();
```

not everyone likes my style though :) ... however if you do want the efficiency savings of line 211 then it should live before line 208